File: nf-inject.js
// lb-inject — runtime bytecode injection for LiquidBounce GraalJS scripts.
//
//   load("/abs/path/nf-inject.js");           // defines globalThis.Inject
//   var h = Inject.inject("net.minecraft.client.Minecraft", "getFps", "HEAD",
//             function () { /* JS hook — runs at the inject point */ });
//   Inject.remove(h);  Inject.list();  Inject.removeAll();
//
// Positions: HEAD, RETURN, BEFORE_INVOKE, AFTER_INVOKE, BEFORE_FIELD, AFTER_FIELD.
// The *_INVOKE/_FIELD positions take a 5th arg = target "owner.member"
// (e.g. "net.minecraft.client.Minecraft.getFps").
//
// hook: a JS function OR a java.lang.Runnable, run at the injection point. A JS
// function runs on whatever thread the patched method runs on, so it's only safe
// for points on the client/render thread (ticks, render, getFps, …). For points
// that fire on other threads, pass a precompiled java.lang.Runnable instead.
//
// Instrumentation is obtained with NO JDK needed at runtime, via the precompiled
// nf-inject-agent.jar (ASM bundled). The library auto-detects how to load it:
//   1. launched with `-javaagent:nf-inject-agent.jar`  -> premain already ran
//      (works on any JRE; nothing else needed)
//   2. else a JDK runtime (jdk.attach present, e.g. GraalVM)  -> the lib spawns
//      the bundled attacher to attach + loadAgent at runtime
//   3. neither -> throws with guidance.
// The injected bytecode is bootstrap-only:
//   ((Runnable) System.getProperties().get("nf.hook.<id>")).run()

(function () {
    const System_ = Java.type("java.lang.System");
    const ProcessBuilder = Java.type("java.lang.ProcessBuilder");
    const ProcessHandle = Java.type("java.lang.ProcessHandle");
    const JString = Java.type("java.lang.String");
    const Paths = Java.type("java.nio.file.Paths");
    const Files = Java.type("java.nio.file.Files");
    const RunnableAdapter = Java.extend(Java.type("java.lang.Runnable"));

    const Integer_ = Java.type("java.lang.Integer");

    function rootFolder() {
        try { return "" + Client.configSystem.rootFolder.getAbsolutePath(); }
        catch (e) { return "" + System_.getProperty("user.dir"); }
    }

    // NfHolder is bootstrap-loaded once the agent is active; before that, Java.type
    // throws (class not found), which we treat as "not loaded yet".
    function holder() { try { return Java.type("NfHolder"); } catch (e) { return null; } }

    const Inject = {
        // Path to the precompiled generic agent jar (nf-holder.jar must sit next to
        // it). Defaults to <LiquidBounce>/scripts/nf-inject-agent.jar.
        agentJar: null,
        _handles: {},
        _n: 0,

        _jar() { return this.agentJar ? ("" + this.agentJar) : Paths.get(rootFolder(), "scripts", "nf-inject-agent.jar").toString(); },

        ready() { const H = holder(); return H !== null && H.injector !== null; },

        // Ensure NfHolder.injector / .inst are available (idempotent).
        ensure() {
            if (this.ready()) return;                                   // -javaagent premain, or prior attach
            const jar = this._jar();
            if (!Files.exists(Paths.get(jar))) throw new Error("nf-inject: agent jar not found at " + jar + " (set Inject.agentJar)");
            // JDK path: spawn the bundled external attacher (needs jdk.attach in java.home)
            const javaBin = "" + System_.getProperty("java.home") + "/bin/java";
            const pid = "" + ProcessHandle.current().pid();
            const pb = new ProcessBuilder(Java.to([javaBin, "-cp", jar, "--add-modules", "jdk.attach", "NfAttacher", pid, jar], "java.lang.String[]"));
            pb.redirectErrorStream(true);
            const proc = pb.start();
            const out = "" + new JString(proc.getInputStream().readAllBytes());
            proc.waitFor();
            if (!this.ready()) {
                throw new Error("nf-inject: could not obtain Instrumentation. Launch with " +
                    "-javaagent:" + jar + " (works on any JRE), or use a JDK runtime (jdk.attach) " +
                    "so the attacher can attach. Attacher said: " + out.trim());
            }
        },

        inject(className, method, position, hook, invokeTarget) {
            this.ensure();
            const H = holder();
            const id = ++this._n;
            const runnable = (typeof hook === "function") ? new RunnableAdapter({ run: hook }) : hook;
            H.hooks.put(Integer_.valueOf(id), runnable);
            const internal = ("" + className).replace(/\./g, "/");
            let tOwner = null, tName = null;
            if (invokeTarget) {
                const s = ("" + invokeTarget).replace(/\//g, ".");
                const dot = s.lastIndexOf(".");
                tOwner = s.slice(0, dot).replace(/\./g, "/");
                tName = s.slice(dot + 1);
            }
            const tr = H.injector.apply(Java.to([H.inst, internal, method, position, id, tOwner, tName], "java.lang.Object[]"));
            const handle = "inj#" + id;
            this._handles[handle] = { tr, id, internal };
            return handle;
        },

        remove(handle) {
            const h = this._handles[handle];
            if (!h) return "no such handle: " + handle;
            const H = holder();
            H.remover.apply(Java.to([H.inst, h.tr, h.internal], "java.lang.Object[]"));
            H.hooks.remove(Integer_.valueOf(h.id));
            delete this._handles[handle];
            return "removed " + handle;
        },
        removeAll() { const ks = Object.keys(this._handles); ks.forEach((k) => this.remove(k)); return "removed " + ks.length; },
        list() { return Object.keys(this._handles); },
    };

    globalThis.Inject = Inject;
})();

File: src/NfInject.java
import java.lang.instrument.ClassFileTransformer;
import java.lang.instrument.Instrumentation;
import java.security.ProtectionDomain;
import java.util.Properties;
import java.util.function.Function;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;

/**
 * Generic, precompiled injection agent for LiquidBounce GraalJS scripts.
 *
 * Two ways it gets an {@link Instrumentation}, both routed through the same
 * static init:
 *   - launch-time:  java -javaagent:nf-inject-agent.jar   -> premain(..)
 *   - runtime:      VirtualMachine.attach(pid).loadAgent(jar) -> agentmain(..)
 *
 * On init it publishes, in System properties:
 *   "nf.inst"            the Instrumentation
 *   "nf.inject.injector" Function<Object[],Object>  apply([inst, classInternal,
 *                          method, position, hookId, targetOwner|null,
 *                          targetName|null]) -> the registered ClassFileTransformer
 *   "nf.inject.remover"  Function<Object[],Object>  apply([inst, transformer,
 *                          classInternal]) -> "removed"
 *   "nf.inject.ready"    "true"
 *
 * The script side puts its hook Runnable at  "nf.hook.<id>"  and the injected
 * bytecode is BOOTSTRAP-ONLY — it does
 *     ((Runnable) System.getProperties().get("nf.hook.<id>")).run()
 * so the target class (loaded by Fabric's Knot loader) needs to resolve nothing
 * but JDK classes. ASM is bundled into this jar, so the injector itself works no
 * matter which classloader loads the agent.
 *
 * Positions: HEAD, RETURN, BEFORE_INVOKE, AFTER_INVOKE, BEFORE_FIELD, AFTER_FIELD.
 */
public final class NfInject {

    public static void premain(String args, Instrumentation inst) { init(inst); }
    public static void agentmain(String args, Instrumentation inst) { init(inst); }

    private static void init(Instrumentation inst) {
        Properties p = System.getProperties();
        p.put("nf.inst", inst);
        p.put("nf.inject.injector", (Function<Object[], Object>) NfInject::inject);
        p.put("nf.inject.remover", (Function<Object[], Object>) NfInject::remove);
        p.put("nf.inject.ready", "true");
    }

    private static Object inject(Object[] a) {
        final Instrumentation inst = (Instrumentation) a[0];
        final String clazz = (String) a[1];          // internal name: net/minecraft/client/Minecraft
        final String method = (String) a[2];
        final String pos = (String) a[3];
        final int id = ((Number) a[4]).intValue();
        final String tOwner = (String) a[5];          // INVOKE/FIELD target owner (internal) or null
        final String tName = (String) a[6];           // INVOKE/FIELD target member or null
        final String key = "nf.hook." + id;

        ClassFileTransformer tr = new ClassFileTransformer() {
            @Override
            public byte[] transform(ClassLoader loader, String cn, Class<?> cbr, ProtectionDomain pd, byte[] buf) {
                if (cn == null || !cn.equals(clazz)) return null;
                ClassReader cr = new ClassReader(buf);
                ClassWriter cw = new ClassWriter(cr, ClassWriter.COMPUTE_MAXS);
                cr.accept(new ClassVisitor(Opcodes.ASM9, cw) {
                    @Override
                    public MethodVisitor visitMethod(int ac, String n, String d, String s, String[] e) {
                        MethodVisitor mv = super.visitMethod(ac, n, d, s, e);
                        if (mv == null || !n.equals(method)) return mv;
                        return new MethodVisitor(Opcodes.ASM9, mv) {
                            // ((Runnable) System.getProperties().get(key)).run()  — bootstrap-only
                            private void fire() {
                                super.visitMethodInsn(Opcodes.INVOKESTATIC, "java/lang/System", "getProperties", "()Ljava/util/Properties;", false);
                                super.visitLdcInsn(key);
                                super.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "java/util/Properties", "get", "(Ljava/lang/Object;)Ljava/lang/Object;", false);
                                super.visitTypeInsn(Opcodes.CHECKCAST, "java/lang/Runnable");
                                super.visitMethodInsn(Opcodes.INVOKEINTERFACE, "java/lang/Runnable", "run", "()V", true);
                            }
                            @Override public void visitCode() {
                                super.visitCode();
                                if (pos.equals("HEAD")) fire();
                            }
                            @Override public void visitInsn(int op) {
                                if (op >= Opcodes.IRETURN && op <= Opcodes.RETURN && pos.equals("RETURN")) fire();
                                super.visitInsn(op);
                            }
                            @Override public void visitMethodInsn(int op, String o, String nm, String dd, boolean itf) {
                                boolean m = tOwner != null && o.equals(tOwner) && nm.equals(tName);
                                if (m && pos.equals("BEFORE_INVOKE")) fire();
                                super.visitMethodInsn(op, o, nm, dd, itf);
                                if (m && pos.equals("AFTER_INVOKE")) fire();
                            }
                            @Override public void visitFieldInsn(int op, String o, String nm, String dd) {
                                boolean m = tOwner != null && o.equals(tOwner) && nm.equals(tName);
                                if (m && pos.equals("BEFORE_FIELD")) fire();
                                super.visitFieldInsn(op, o, nm, dd);
                                if (m && pos.equals("AFTER_FIELD")) fire();
                            }
                        };
                    }
                }, 0);
                return cw.toByteArray();
            }
        };
        inst.addTransformer(tr, true);
        try {
            inst.retransformClasses(Class.forName(clazz.replace('/', '.'), false, Thread.currentThread().getContextClassLoader()));
        } catch (Throwable ex) {
            inst.removeTransformer(tr);
            throw new RuntimeException(String.valueOf(ex));
        }
        return tr;
    }

    private static Object remove(Object[] a) {
        Instrumentation inst = (Instrumentation) a[0];
        inst.removeTransformer((ClassFileTransformer) a[1]);
        try {
            inst.retransformClasses(Class.forName(((String) a[2]).replace('/', '.'), false, Thread.currentThread().getContextClassLoader()));
        } catch (Throwable ignored) { }
        return "removed";
    }

    private NfInject() { }
}
